### PR TITLE
Fixed memory leak in HcalDigitizer

### PR DESCRIPTION
**Pull request by Dr15Jones**
valgrind had found a memory leak in HcalDigitizer. Changing the member data to be held by value or via std::unique_ptr solved the problem.

### CODE REVIEW

File: SimCalorimetry/HcalSimProducers/src/HcalDigitizer.cc [hunk 1]
```diff
@@ -450,14 +417,14 @@ void HcalDigitizer::finalizeEvent(edm::Event& e, const edm::EventSetup& eventSet
...
       !theHFQIE10DetIds.empty() ? 
-      theParameterMap->simParameters(theHFQIE10DetIds[0]).readoutFrameSize() : 
+      theParameterMap.simParameters(theHFQIE10DetIds[0]).readoutFrameSize() : 
       QIE10DigiCollection::MAXSAMPLES
     )
   );
   std::unique_ptr<QIE11DigiCollection> hbheQIE11Result(
     new QIE11DigiCollection(
       !theHBHEQIE11DetIds.empty() ? 
-      ((HcalSiPMHitResponse *)theHBHESiPMResponse)->getReadoutFrameSize(theHBHEQIE11DetIds[0]) :
```
Comment (Dr15Jones) [inline SimCalorimetry/HcalSimProducers/src/HcalDigitizer.cc]: I avoided the cast by changing the variable to be the correct type

File: SimCalorimetry/HcalSimProducers/src/HcalDigitizer.cc [hunk 2]
```diff
@@ -604,12 +571,12 @@ void HcalDigitizer::updateGeometry(const edm::EventSetup & eventSetup) {
   //handle mixed QIE8/11 scenario in HBHE
   buildHBHEQIECells(hbheCells,eventSetup);
   if(theHBHESiPMResponse)
-    ((HcalSiPMHitResponse *)theHBHESiPMResponse)->setDetIds(theHBHEQIE11DetIds);
```
Comment (Dr15Jones) [inline SimCalorimetry/HcalSimProducers/src/HcalDigitizer.cc]: Avoided the cast by changing the variable to be the correct type.

File: SimCalorimetry/HcalSimProducers/src/HcalDigitizer.cc [hunk 3]
```diff
@@ -604,12 +571,12 @@ void HcalDigitizer::updateGeometry(const edm::EventSetup & eventSetup) {
...
+    theHBHESiPMResponse->setDetIds(theHBHEQIE11DetIds);
   
   if(theHOSiPMDigitizer) {
     buildHOSiPMCells(hoCells, eventSetup);
     if(theHOSiPMResponse)
-      ((HcalSiPMHitResponse *)theHOSiPMResponse)->setDetIds(hoCells);
```
Comment (Dr15Jones) [inline SimCalorimetry/HcalSimProducers/src/HcalDigitizer.cc]: Avoided the cast by changing the variable to be the correct type.

**Comment by Dr15Jones** [comment]
please test

**Comment by civanch** [comment]
+1

**Pull request merged**